apso/src/main/scala/eu/shiftforward/apso/io/InsistentInputStream.scala

```diff
@@ -36,21 +38,37 @@ class InsistentInputStream(streamBuilder: () => InputStream, maxRetries: Int = 1
   }
 
   @tailrec
-  private[this] def readRetries(remainingTries: Int): Int = {
-    Try(innerStream.read) match {
+  private[this] def readRetries(remainingTries: Int, f: => Int): Int = {
+    Try(f) match {
       case Success(n) =>
-        currPos += 1
         n
       case Failure(t) =>
         if (remainingTries <= 0) throw t
         else {
+          log.warn("Failed to read from stream: {}", t)
           val nextRemainingTries = retryStreamCreation(remainingTries - 1)
-          readRetries(nextRemainingTries)
+          readRetries(nextRemainingTries, f)
         }
     }
   }
 
-  def read(): Int = readRetries(maxRetries)
+  def read(): Int = {
+    val nextByte = readRetries(maxRetries, innerStream.read())
+    if (nextByte > 0) currPos += 1
```

Comment:
You are correct (I messed up when I was copying code from the other implementations).

Fixed in 04e1b50.